students/andrew_garcia/lesson07-assignment/parallel.py

```diff
@@ -0,0 +1,212 @@
+                            'address': item['address'],
+                            'phone_number': item['phone_number'], 'email': item['email']}
+                LOGGER.info('Rentals Added')
+                rentals.insert_one(csv_item)
+                added_rentals += 1
+
+    except FileNotFoundError:
+        LOGGER.error('File Not Found.')
+        print('File %s Does Not Exist', rental_file)
+
+    rental_end_time = time.time()
+    rental_time = rental_end_time - rental_start_time
+    rental_count_end = rentals.count_documents({})
+    rental_details = (added_rentals, rental_count_start, rental_count_end, rental_time)
+    return rental_details
+
+
+def show_available_products():
+    """ Returns dictionary of products listed as available with product id, description,
+    product type, quantity available, """
+    mongo = MongodbConnection()
+
+    with mongo:
+        LOGGER.info('Connecting to Database')
+        db = mongo.connection.media
+        available_products = {}
+        LOGGER.info('Finding Available Products')
+        for item in db.products.find():
+            product_info = {'product_id': item['product_id'], 'description': item['description'],
+                            'product_type': item['product_type'],
+                            'quantity_available': item['quantity_available']}
+            available_products[item['product_id']] = product_info
+        return available_products
+
+
+def show_rentals(product_id):
+    """ Returns dictionary with user id, name, address, phone number, and email who are renting a
+    certain product"""
+    mongo = MongodbConnection()
+
+    with mongo:
+        LOGGER.info('Connecting to Database')
+        db = mongo.connection.media
+        rental_data = {}
+        LOGGER.info('Finding Rentals for Product')
+        for item in db.rentals.find({'product_id': product_id}):
+            rental_info = {'name': item['name'], 'rental_id': item['rental_id'],
+                           'address': item['address'], 'phone_number': item['phone_number'],
+                           'email': item['email']}
+            rental_data[item['user_id']] = rental_info
+        return rental_data
+
+
+def clear_data():
+    """ Clears Entire Database """
+    mongo = MongodbConnection()
+
+    with mongo:
+        LOGGER.info('Connecting to Database')
+        db = mongo.connection.media
+        db.products.drop()
+        db.customers.drop()
+        db.rentals.drop()
+
+
+if __name__ == '__main__':
+    start_time = time.time()
+    path = 'D:\\UW\\PY220\\Examples and Assignments\\lesson07\\assignment\\csv_files'
+    threads = []
+
+    threads.append(threading.Thread(target=import_products, args=(path, 'products.csv')))
+    threads.append(threading.Thread(target=import_customers, args=(path, 'customers.csv')))
+    threads.append(threading.Thread(target=import_rentals, args=(path, 'rentals.csv')))
+
+    for thread in threads:
+        thread.start()
+
```

Comment:
Hi Andrew,

The implementation of threading looks good, but you're missing one step: Joining.  The three lines of code you have after _thread.start()_ on line 208 will run as soon as the last thread is started and you exit the for loop.

You have conveniently packed all three threads into a single list, _threads_, you can now add another for loop to do the joining:

```
    for thread in threads:
        thread.join()
```

This works as a "meeting point" of sorts for all three threads, to make sure each has finished before (and all data has been processed) before concluding execution. Note that the size of the data set may not be big enough to cause a significant difference between linear and parallel implementations, in particular if you have a relatively new (and fast) system.

Also, remember to run Pylint on _parallel.py_ and fix any outstanding issues.

Let me know if you have any additional questions.

Happy New Year!

-Luis